runelite-client/src/main/java/net/runelite/client/graphics/ModelOutlineRenderer.java

```diff
@@ -0,0 +1,1073 @@
+import net.runelite.api.NPC;
+import net.runelite.api.NPCComposition;
+import net.runelite.api.Perspective;
+import net.runelite.api.Player;
+import net.runelite.api.Projectile;
+import net.runelite.api.coords.LocalPoint;
+import net.runelite.client.task.Schedule;
+import net.runelite.client.task.Scheduler;
+
+@Singleton
+public class ModelOutlineRenderer
+{
+	/*
+	 * This class doesn't really "need" static variables, but they are
+	 * static for performance reasons. Arrays are kept outside methods
+	 * to avoid frequent big allocations. Arrays should mostly be seen
+	 * as ArrayLists. The size of them is increased whenever they need
+	 * to become bigger.
+	 */
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private Scheduler scheduler;
+
+	private boolean isReset;
+	private boolean usedSinceLastCheck;
+
+	// Dimensions of the underlying image
+	private int imageWidth;
+	private int imageHeight;
+
+	// Boundaries for the current rasterization
+	private int clipX1;
+	private int clipY1;
+	private int clipX2;
+	private int clipY2;
+
+	// Pixel points that would be rendered to
+	private int[] visited;
+	private int currentVisitedNumber = 0;
+
+	// Transformed vertex positions
+	private int[] projectedVerticesX;
+	private int[] projectedVerticesY;
+	private boolean[] projectedVerticesRenderable;
+
+	// An array of pixel points to raster onto the image. These are checked against
+	// clip boundaries and the visited array to prevent drawing on top of the model
+	// and outside the scene area. They are grouped per distance to the closest pixel
+	// drawn on the model.
+	private int[][] outlinePixels;
+	private int[] outlinePixelsLengths; // outlinePixelsLength[i] is the used length of outlinePixels[i]
+	private int outlineArrayWidth;
+
+	// A list of pixel distances ordered from shortest to longest distance for
+	// each outline width. These are calculated once upon first usage and then
+	// stored here to prevent reevaluation.
+	private List<List<PixelDistanceAlpha>> precomputedDistancePriorities;
+
+	public ModelOutlineRenderer()
+	{
+		reset();
+	}
+
+	@Schedule(period = 5, unit = ChronoUnit.SECONDS)
+	public void checkUsage()
+	{
+		if (!isReset && !usedSinceLastCheck)
+		{
+			// Reset memory allocated when the rasterizer becomes inactive
+			reset();
+		}
+		usedSinceLastCheck = false;
+	}
+
+	/**
+	 * Reset memory used by the rasterizer
+	 */
+	private void reset()
+	{
+		visited = new int[0];
+		projectedVerticesX = new int[0];
+		projectedVerticesY = new int[0];
+		projectedVerticesRenderable = new boolean[0];
+		outlinePixels = new int[0][];
+		outlinePixelsLengths = new int[0];
+		precomputedDistancePriorities = new ArrayList<>(0);
+		isReset = true;
+
+		if (scheduler != null)
+		{
+			scheduler.unregisterObject(this);
+		}
+	}
+
+	/**
+	 * Calculate the next power of two of a value
+	 *
+	 * @param value The value to find the next power of two of
+	 * @return Returns the next power of two
+	 */
+	private int nextPowerOfTwo(int value)
```

Comment:
Most methods in this class can still be static (or well at least those that do not depend on state)